### ISSUE TITLE
Raspberrypi 400: eth0: failed to connect to PHY

### ISSUE THREAD

**Issue by avdv** [NixOS/nixpkgs#183019]
### Describe the bug

After booting the network device cannot be used. The device driver bcmgenet is not properly initialized.

### Steps To Reproduce
1. download an SD card image, e.g. `nixos-sd-image-22.05.1744.07a2e6a4e31-aarch64-linux.img.zst`
2. write to a SD card
3. boot using the SD card

### Expected behavior

Usable eth0 interface.

### Screenshots

![image](https://user-images.githubusercontent.com/3471749/181103584-e50c05b5-6c53-44f3-be1a-616482907951.png)

### Additional context

I tried images for NixOS 22.05 as well as the latest from trunk:

* `nixos-sd-image-22.05.1744.07a2e6a4e31-aarch64-linux.img.zst`
* `nixos-sd-image-22.11pre393585.42ca9bef09e-aarch64-linux.img.zst`

dmesg output:
```
[Tue Jul 26 21:00:54 2022] Linux version 5.15.32 (nixbld@localhost) (gcc (GCC) 9.3.0, GNU ld (GNU Binutils) 2.38) #1-NixOS SMP Tue Jan 1 00:00:00 UTC 1980
[Tue Jul 26 21:00:54 2022] Machine model: Raspberry Pi 400 Rev 1.1
[Tue Jul 26 21:00:55 2022] bcmgenet fd580000.ethernet: GENET 5.0 EPHY: 0x0000
[Tue Jul 26 21:00:55 2022] mdio_bus unimac-mdio--19: MDIO device at address 1 is missing.
[Tue Jul 26 21:00:55 2022] unimac-mdio unimac-mdio.-19: Broadcom UniMAC MDIO bus
```

Using the official Raspberry Pi OS, this works seamlessly.

I also tried to use different firmware (copied from the working Raspberry Pi OS), and updated the eeprom to no avail.

Next, I tried to use different dtbs but made a mistake while doing so... This lead to an error that the `bcm2711-rpi-400.dtb` file could not be found by the boot loader. BUT that actually works! :tada: 

![image](https://user-images.githubusercontent.com/3471749/181104407-4e9ac1df-67fc-49fe-9196-01d32391cdf9.png)

dmesg output from the successful boot:
```
[Tue Jul 26 21:04:05 2022] unimac-mdio unimac-mdio.-19: Broadcom UniMAC MDIO bus
[Tue Jul 26 21:04:05 2022] bcmgenet fd580000.ethernet: GENET 5.0 EPHY: 0x0000
[Tue Jul 26 21:04:05 2022] unimac-mdio unimac-mdio.-19: Broadcom UniMAC MDIO bus
[Tue Jul 26 21:04:10 2022] bcmgenet fd580000.ethernet: configuring instance for external RGMII (RX delay)
[Tue Jul 26 21:04:10 2022] bcmgenet fd580000.ethernet eth0: Link is Down
[Tue Jul 26 21:04:14 2022] bcmgenet fd580000.ethernet eth0: Link is Up - 1Gbps/Full - flow control off
```
### Notify maintainers

@bartsch @dezgeg @samueldr @lopsided98 (might be an u-boot issue?)

### Metadata

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"aarch64-linux"`
 - host os: `Linux 5.15.32, NixOS, 22.05 (Quokka), 22.05.20220627.62de360`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.8.1`
 - channels(root): `""`
 - nixpkgs: `/etc/nixpkgs`

```

**Comment by avdv** [comment]
Setting `boot.loader.generic-extlinux-compatible.useGenerationDeviceTree = false` causes the `FDTDIR` entries in `extlinux.conf` to vanish and after a reboot the eth0 device is working.

AFAIK, the `start4.elf` second stage boot loader already loads the relevant dtb and passes it on to u-boot. u-boot - via the `FDTDIR` setting - is also loading a dtb which seems to cause hiccup.

I tried to use the )same_ dtb in both places (copied `bcm2711-rpi-400.dtb` from the `FIRMWARE` partition to `/boot/extlinux/dtbs/broadcom/` on the `NIXOS_SD` partition and modified `extlinux.conf` by setting `FDTDIR` accordingly (as you can see in the screenshot above)  but that did not help. It only works if I don't load a dtb from u-boot at all.

**Comment by karthanistyr** [comment]
I'm interested in seeing this fixed, as I'm hitting it too on a Pi 400. I've searched other occurrences of this problem and came across this: procount/pinn/issues/566
They seem to have fixed it somehow. Would any info there be relevant to the issue here ?

**Comment by avdv** [comment]
Thanks for the pointer. I've read through that issue and as far as I understand it, the OP of the PINN issue was missing the `CONFIG_BROADCOM_PHY` kernel module and adding it fixed it for them.

I am adding a bit more data points for my system:

`Model		: Raspberry Pi 400 Rev 1.1`
```console
$ sudo mii-tool -v eth0
eth0: negotiated 1000baseT-FD flow-control, link ok
  product info: vendor 00:08:85, model 36 rev 2
  basic mode:   autonegotiation enabled
  basic status: autonegotiation complete, link ok
  capabilities: 1000baseT-FD 100baseTx-FD 100baseTx-HD 10baseT-FD 10baseT-HD
  advertising:  1000baseT-FD 100baseTx-FD 100baseTx-HD 10baseT-FD 10baseT-HD flow-control
  link partner: 1000baseT-FD 100baseTx-FD 100baseTx-HD 10baseT-FD 10baseT-HD
```

```console
$ zgrep BROADCOM_PHY /proc/config.gz 
CONFIG_BROADCOM_PHY=y
```

**Comment by karthanistyr** [comment]
I could boot a rpi 4b without issues (and with ethernet working) using the same image, so I guess the problem is pi 400 specific...
I think this is the important error message: `[    1.405791] mdio_bus unimac-mdio--19: MDIO device at address 1 is missing.
`
I'm trying to use a variation of firmwares and kernels but my noobness in nix is impeding my progress

**Comment by karthanistyr** [comment]
per request: https://gist.github.com/karthanistyr/f9ca1b9e49277853e044aee97dfb1bc4

Diff of the running rpi OS device tree and the firmware-supplied pi 400 dtb
Raw dts for both in the sale gist

**Comment by karthanistyr** [comment]
more context: diff between a working rpi OS live device tree and a non-working nixos live device tree
https://gist.github.com/karthanistyr/c6ccfb85f1faea4aa012a442ec0b0040

with the kind help of @K900 and @samueldr 

relevant line: https://gist.github.com/karthanistyr/c6ccfb85f1faea4aa012a442ec0b0040#file-live_device_trees-diff-L162
```diff
@@ -626,7 +662,7 @@
 				ethernet-phy@1 {
 					led-modes = <0x00 0x08>;
 					phandle = <0x2e>;
-					reg = <0x01>;
+					reg = <0x00>;
 				};
 			};
 		};
```

I quickly hacked together the firmware-provided `bcm2711-rpi-400.dtb` and changed the reg value for `ethernet-phy@1` from `0x01` to `0x00`, recompiled and overwrote both dtbs on the sd card in `FIRMWARE` and `NIXOS_SD/boot/nixos/<whatever>-dtbs/broadcom`.

ethernet comes up and works

**Comment by tobiasBora** [comment]
I can't help (sorry to ask my question here) but by any chance do you have some references on how to get the device trees and diff them?

**Comment by spacecowb0y** [comment]
I can confirm the issue is still open.

**Comment by K900** [comment]
The uboot release with the fix was two days ago, we still need to update the version in nixpkgs.

**Comment by K900** [comment]
(nixpkgs update PR: https://github.com/NixOS/nixpkgs/pull/227947)

**Comment by spacecowb0y** [comment]
@K900 sweet. Thanks for the headsup, gonna test it asap.

**Comment by avdv** [comment]
I finally upgraded my raspberry pi 400 to nixos-24.05 recently. And I noticed sound was not working. That was because I still had the workaround of `boot.loader.generic-extlinux-compatible.useGenerationDeviceTree = false` in place.

When I removed that setting, sound was working, but the eth0 PHY problem was back.

After upgrading u-boot on the SD (by overwriting the `u-boot-rpi4.bin` on the FIRMWARE partition with a fresh one from nixpkgs) both, audio and network, are up and running.
